tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,553 @@
+    })) {
+      return false;
+    }
+
+    return MergeClusters(from, to);
+  }
+
+  bool RunEdgeContractionLoop() {
+    bool changed = false;
+    using namespace std::placeholders;
+    changed = ForEachEdgeInPostOrder(
+        std::bind(&FusionPlanner::TryToContractEdge, this, _1, _2));
+    return changed;
+  }
+
+  SmallVector<Operation*, 4> op_list_;
+
+  // Shape equality checker
+  ShapeConstraintAnalysis shape_analysis_;
+
+  std::unordered_map<Operation*, int> op_to_node_id_;
+
+  GraphCycles cycle_detector_;
+  std::vector<std::unique_ptr<Cluster>> cluster_storage_;
+  std::vector<UnionFind<Cluster*>> cluster_for_node_;
+};
+
+struct HloOutlineFusion : public ModulePass<HloOutlineFusion> {
+  void runOnModule() override {
+    SmallVector<FuncOp, 4> funcs;
+    // We may add some fusion func into the module during
+    // visiting, thus we collect all target funcs first.
+    for (FuncOp func : getModule().getOps<FuncOp>()) {
+      if (IsTargetFunc(func)) {
+        funcs.push_back(func);
+      }
+    }
+    for (FuncOp func : funcs) {
+      runOnFunction(func);
+    }
+  }
+
+  bool IsTargetFunc(FuncOp func) {
+    // We currently only support single block function.
+    if (func.getBlocks().size() != 1) {
+      return false;
+    }
+
+    auto ctx = func.getContext();
+    Dialect *hlo_dialect = ctx->getRegisteredDialect("xla_hlo");
+    Dialect *lhlo_dialect = ctx->getRegisteredDialect("xla_lhlo");
+    if (llvm::none_of(func.getBlocks().front(), [&](Operation& op) {
+        return (op.getDialect() == hlo_dialect ||
+                op.getDialect() == lhlo_dialect);
+    })) {
+      // no target ops
+      return false;
+    }
+
+    // not handle already-fused func
+    if (func.getAttrOfType<UnitAttr>(kHloTensorFuncAttr)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  void runOnFunction(FuncOp func) {
+    SmallVector<Operation*, 4> op_list;
+    for (auto& op : func.getBlocks().front()) {
+      op_list.push_back(&op);
+    }
+    FusionPlanner planner(op_list);
+    auto plan = planner.Run();
+    if (!plan) {
+      emitError(func.getLoc(), "Fusion planner failed.");
+      signalPassFailure();
+      return;
+    }
+    if (!ApplyFusionPlan(func, *plan)) {
+      emitError(func.getLoc(), "Apply Fusion Plan failed.");
+      signalPassFailure();
+      return;
+    }
+  }
+
+  bool ApplyFusionPlan(FuncOp func, const FusionPlan& plan) {
+    ModuleOp module = getModule();
+    Builder builder(module.getContext());
+    SymbolTable manager(module);
+
+    for (auto& pattern: plan) {
+      SmallVector<Value, 4> inputs = GetInputsOfFusionPattern(pattern);
+      SmallVector<Value, 4> outputs = GetOutputsOfFusionPattern(pattern);
+      SmallVector<Type, 4> input_types;
+      SmallVector<Type, 4> output_types;
+      input_types.reserve(inputs.size());
+      output_types.reserve(outputs.size());
+      for(auto& v : inputs) input_types.push_back(v.getType());
```

Comment:
Avoid one line loop to prevent from potential typos......
for () ....;
|
|
->
    for () ... ; ...; 
The newly added stmt will be missed from the for loop....

The

Reply:
Clang-format would catch this
(one line statement are common in MLIR code FYI)